yamui-screensaverd.c

```diff
@@ -0,0 +1,361 @@
+}
+
+/* ------------------------------------------------------------------------ */
+
+/* Open all /dev/input/event* files. */
+static int
+open_fds(int fds[], int *num, int max_num)
+{
+	DIR *dir;
+	struct dirent *d;
+
+	if (!(dir = opendir(DEV_INPUT_DIR))) {
+		errorf("Can't open directory %s", DEV_INPUT_DIR);
+		return -1;
+	}
+
+	while ((d = readdir(dir))) {
+		char name[256];
+
+		if (strncmp(d->d_name, EVENT_PREFIX, strlen(EVENT_PREFIX)))
+			continue; /* Not /dev/input/event* file */
+
+		snprintf(name, sizeof(name), "%s/%s", DEV_INPUT_DIR,
+			 d->d_name);
+		debugf("Processing input ivents file %s", name);
+		if ((fds[*num] = open(name, O_RDONLY)) == -1) {
+			errorf("Can't open input device %s", name);
+			continue;
+		}
+
+		if (check_device_type(fds[*num]) == -1) {
+			close(fds[*num]);
+			continue;
+		}
+
+		if (++*num >= max_num)
+			break;
+	}
+
+	closedir(dir);
+	if (!*num) {
+		infof("No suitable input events devices found.");
+		return -1;
+	}
+
+	return 0;
+}
+
+/* ------------------------------------------------------------------------ */
+
+static int
+sysfs_write_int(const char *fname, int val)
+{
+	FILE *f;
+
+	if (!(f = fopen(fname, "w"))) {
+		errorf("Can't open \"%s\" for writing", fname);
+		return -1;
+	}
+
+	fprintf(f, "%d\n", val);
+	fclose(f);
+	return 0;
+}
+
+/* ------------------------------------------------------------------------ */
+
+static int is_display_on = 1;
+
+static int
+turn_display_on(void)
+{
+	if (is_display_on)
+		return 0;
+
+	debugf("Turning display on.");
+	is_display_on = 1;
+	return sysfs_write_int(DISPLAY_CONTROL, 0);
+}
+
+/* ------------------------------------------------------------------------ */
+
+static int
+turn_display_off(void)
+{
+	if (!is_display_on)
+		return 0;
+
+	debugf("Turning display off.");
+	is_display_on = 0;
+	return sysfs_write_int(DISPLAY_CONTROL, 1);
+}
+
+/* ------------------------------------------------------------------------ */
+
+static int
+handle_events(int fd)
+{
+	int rv;
+	char buf[EVENTS_BUF_SIZE * sizeof(struct input_event)];
+
+	/* Read and ignore event data if OK. */
+	rv = read(fd, buf, sizeof(buf));
+	if (rv < 0) {
+		errorf("Error on read");
+		return -1;
+	} else if (rv == 0) {
+		infof("Unexpected EOF on read");
+		return -1;
+	}
+
+	return 0;
+}
+
+/* ------------------------------------------------------------------------ */
+
+void
+signal_handler(int sig UNUSED)
+{
+	running = 0;
+}
+
+/* ------------------------------------------------------------------------ */
+
+int
+main(void)
+{
+	int fds[MAX_DEVICES], num_fds = 0, ret = EXIT_SUCCESS;
+
+	if (open_fds(fds, &num_fds, MAX_DEVICES) == -1)
+		return EXIT_FAILURE;
+
+	debugf("Started");
+	signal(SIGINT,  signal_handler);
+	signal(SIGTERM, signal_handler);
+
+	while (running) { /* Main loop */
+		int i, rv, max_fd = 0;
+		fd_set rfds;
+		struct timeval tv;
+
+		FD_ZERO(&rfds);
+		for (i = 0; i < num_fds; i++) {
+			FD_SET(fds[i], &rfds);
+			max_fd = fds[i] > max_fd ? fds[i] : max_fd;
```

Comment:
Feel free to ignore, but that ternary operator feels like more complex way to say:

    if (max_fd < fds[i]) max_fd = fds[i];

Also, since the fd array does not change during runtime, the max fd could be scanned just once before entering the running loop.